# DataFrame

It's heart of pandas libary. DataFrame is a two dimensional series object which has index and multiple columns. DataFrame is used for Data Cleaning and Data Analysis. We can define DataFrame as 2-Axis Value Array

# DataFrame using Series

In [2]:
#Exercise
import pandas as pd

record1 = pd.Series({"Name": "Alice",
                     "Subject": "Chemistry",
                     "Marks": 86})

record2 = pd.Series({"Name": "John",
                     "Subject": "Chemistry",
                     "Marks": 90})



record3 = pd.Series({"Name": "Jenny",
                     "Subject": "Math",
                     "Marks": 95})

In [3]:
df = pd.DataFrame([record1, record2, record3], index =["school1", "school2", "school1"])
df.head(3)

Name    Subject  Marks
school1  Alice  Chemistry     86
school2   John  Chemistry     90
school1  Jenny       Math     95

# DataFrame using Dictionary

In [4]:
students = [{"Name": "John",
                     "Subject": "Chemistry",
                     "Marks": 90},
            {"Name": "Jack",
                     "Subject": "Physics",
                     "Marks": 86},
            {"Name": "Jashmin",
                       "Subject": "Math",
                        "Marks": 95}]

df1 = pd.DataFrame(students, index =['school1', 'school2', 'school3'])
df1.head(5)

Name    Subject  Marks
school1     John  Chemistry     90
school2     Jack    Physics     86
school3  Jashmin       Math     95

# Loc and iLOC in Data Sele

In [8]:
df.iloc[0:2]


Name    Subject  Marks
school1  Alice  Chemistry     86
school2   John  Chemistry     90

In [6]:
df.loc['school1']

Name    Subject  Marks
school1  Alice  Chemistry     86
school1  Jenny       Math     95

In [9]:
type(df.loc['school1'])

pandas.core.frame.DataFrame

In [10]:
type(df.loc['school2'])

pandas.core.series.Series

In [11]:
df.loc['school1', 'Name']

school1    Alice
school1    Jenny
Name: Name, dtype: object

**Transpose** represents as **`T`**. Columns become rows and rows become columns

In [12]:
df.T

school1    school2 school1
Name         Alice       John   Jenny
Subject  Chemistry  Chemistry    Math
Marks           86         90      95

In [10]:
df['Name']

school1    Alice
school2     John
school1    Jenny
Name: Name, dtype: object

In [13]:
df.T.loc['Name']

school1    Alice
school2     John
school1    Jenny
Name: Name, dtype: object

In [12]:
df.loc['school1']['Name']

school1    Alice
school1    Jenny
Name: Name, dtype: object

In [14]:
type(df.loc['school1'])

pandas.core.frame.DataFrame

In [14]:
type(df.loc['school1']['Name'])

pandas.core.series.Series

**`:,`** is used to select all row index values and other is a list of specific col

In [15]:
df.loc[:, ['Name', 'Marks']]

Name  Marks
school1  Alice     86
school2   John     90
school1  Jenny     95

# Drop / Delete Data

In [16]:
df.drop('school1')

Name    Subject  Marks
school2  John  Chemistry     90

As we can see, it will not actually  drop the value from DataFrame or update anything in DataFrame

In [17]:
df

Name    Subject  Marks
school1  Alice  Chemistry     86
school2   John  Chemistry     90
school1  Jenny       Math     95

To update the DataFrame after droping a value, we need to use **`inplace = True`**

To apply drop on column, **`axis=1`**  and To apply drop on row, **`axis=0`**

**`.copy()`** is for copy the DataFrame

In [18]:
df1 = df.copy()
print("before drop:", '\n', df1)


before drop: 
           Name    Subject  Marks
school1  Alice  Chemistry     86
school2   John  Chemistry     90
school1  Jenny       Math     95


In [19]:
df1.drop('Name', axis=1, inplace = True)

print("After drop the col:", '\n', df1)

After drop the col: 
            Subject  Marks
school1  Chemistry     86
school2  Chemistry     90
school1       Math     95


In [20]:
# Delete Column

In [21]:
del df1['Marks']
df1

Subject
school1  Chemistry
school2  Chemistry
school1       Math

# ADD New Column 



In [22]:
df1['New Col']= None
df1

Subject New Col
school1  Chemistry    None
school2  Chemistry    None
school1       Math    None

# Load CSV File and Explore DF
We are gonna use shell command **`cat`** for concate DataFrame for CSV file

In [23]:
!cat resources/week-2/datasets/Admission_Predict.csv

















































































































































































































































































































































































































400,333,117,4,5,4,9.66,1,0.95

In [24]:
import pandas as pd
df = pd.read_csv("resources/week-2/datasets/Admission_Predict.csv")
df.head(5)

Serial No.  GRE Score  TOEFL Score  University Rating  SOP  LOR   CGPA  \
0           1        337          118                  4  4.5   4.5  9.65   
1           2        324          107                  4  4.0   4.5  8.87   
2           3        316          104                  3  3.0   3.5  8.00   
3           4        322          110                  3  3.5   2.5  8.67   
4           5        314          103                  2  2.0   3.0  8.21   

   Research  Chance of Admit   
0         1              0.92  
1         1              0.76  
2         1              0.72  
3         1              0.80  
4         0              0.65

In [24]:
import pandas as pd
df = pd.read_csv("resources/week-2/datasets/Admission_Predict.csv", index_col = 0) #to make row index starts with 1
df.head(3)

GRE Score  TOEFL Score  University Rating  SOP  LOR   CGPA  \
Serial No.                                                               
1                 337          118                  4  4.5   4.5  9.65   
2                 324          107                  4  4.0   4.5  8.87   
3                 316          104                  3  3.0   3.5  8.00   

            Research  Chance of Admit   
Serial No.                              
1                  1              0.92  
2                  1              0.76  
3                  1              0.72

In [25]:
# Column Rename in a DataFrame

In [26]:
new_df = df.rename(columns={'GRE Score':'GRE Score', 'TOFEL Score':'TOFEL Score', 
                           'University Rating':'University Rating', 'SOP': 'Statement of Purpose',
                          'LOR': 'Letter of Recommandation', 'CGPA':'CGPA', 'Research': 'Research', 'Chance of Admit': 'Chance of Admit'})
new_df.head(3)

GRE Score  TOEFL Score  University Rating  Statement of Purpose  \
Serial No.                                                                    
1                 337          118                  4                   4.5   
2                 324          107                  4                   4.0   
3                 316          104                  3                   3.0   

            LOR   CGPA  Research  Chance of Admit   
Serial No.                                          
1            4.5  9.65         1              0.92  
2            4.5  8.87         1              0.76  
3            3.5  8.00         1              0.72

In [28]:
new_df.columns # as there is a space in LOR and Chance of Admit that's why it is not being replaced

Index(['GRE Score', 'TOEFL Score', 'University Rating', 'Statement of Purpose',
       'LOR ', 'CGPA', 'Research', 'Chance of Admit '],
      dtype='object')

In [27]:
new_df = new_df.rename (columns={'LOR ': 'Letter of Recommandation'}) 
new_df.head(3)

GRE Score  TOEFL Score  University Rating  Statement of Purpose  \
Serial No.                                                                    
1                 337          118                  4                   4.5   
2                 324          107                  4                   4.0   
3                 316          104                  3                   3.0   

            Letter of Recommandation  CGPA  Research  Chance of Admit   
Serial No.                                                              
1                                4.5  9.65         1              0.92  
2                                4.5  8.87         1              0.76  
3                                3.5  8.00         1              0.72

# Bestway to Trimmed unnecessary space
**`mapper=str.strip, axis ='columns'`** is to remove unnecessary space applied in column axis

In [28]:
new_df = new_df.rename(mapper=str.strip, axis ='columns') 
new_df.columns

Index(['GRE Score', 'TOEFL Score', 'University Rating', 'Statement of Purpose',
       'Letter of Recommandation', 'CGPA', 'Research', 'Chance of Admit'],
      dtype='object')

Change the column name into lower case

In [29]:
cols = df.columns.tolist()

#use of list comprehension
cols=[x.lower().strip() for x in cols]
cols

['gre score',
 'toefl score',
 'university rating',
 'sop',
 'lor',
 'cgpa',
 'research',
 'chance of admit']

In [23]:
df.columns = cols
df.head(3)

gre score  toefl score  university rating  sop  lor  cgpa  \
Serial No.                                                              
1                 337          118                  4  4.5  4.5  9.65   
2                 324          107                  4  4.0  4.5  8.87   
3                 316          104                  3  3.0  3.5  8.00   

            research  chance of admit  
Serial No.                             
1                  1             0.92  
2                  1             0.76  
3                  1             0.72

# Boolen Mask
one dimension for Series
multi dimension for DataFrame

values in the array True/False

**`index_col=0`** = no index column

In [25]:
import pandas as pd
df = pd.read_csv("resources/week-2/datasets/Admission_Predict.csv", index_col=0)
# clear the space and upper case
df.columns=[x.lower().strip() for x in df.columns]
df.head(5)


gre score  toefl score  university rating  sop  lor  cgpa  \
Serial No.                                                              
1                 337          118                  4  4.5  4.5  9.65   
2                 324          107                  4  4.0  4.5  8.87   
3                 316          104                  3  3.0  3.5  8.00   
4                 322          110                  3  3.5  2.5  8.67   
5                 314          103                  2  2.0  3.0  8.21   

            research  chance of admit  
Serial No.                             
1                  1             0.92  
2                  1             0.76  
3                  1             0.72  
4                  1             0.80  
5                  0             0.65

In [26]:
df['chance of admit'] >0.7

Serial No.
1       True
2       True
3       True
4       True
5      False
       ...  
396     True
397     True
398     True
399    False
400     True
Name: chance of admit, Length: 400, dtype: bool

note = it only returns the boolen series result

# Use of where

to get full details of all rows based on column value

In [27]:
df.where(df['chance of admit']>0.7).head()

gre score  toefl score  university rating  sop  lor  cgpa  \
Serial No.                                                              
1               337.0        118.0                4.0  4.5  4.5  9.65   
2               324.0        107.0                4.0  4.0  4.5  8.87   
3               316.0        104.0                3.0  3.0  3.5  8.00   
4               322.0        110.0                3.0  3.5  2.5  8.67   
5                 NaN          NaN                NaN  NaN  NaN   NaN   

            research  chance of admit  
Serial No.                             
1                1.0             0.92  
2                1.0             0.76  
3                1.0             0.72  
4                1.0             0.80  
5                NaN              NaN

In [28]:
df.where(df['chance of admit']>0.7).dropna().head()

gre score  toefl score  university rating  sop  lor  cgpa  \
Serial No.                                                              
1               337.0        118.0                4.0  4.5  4.5  9.65   
2               324.0        107.0                4.0  4.0  4.5  8.87   
3               316.0        104.0                3.0  3.0  3.5  8.00   
4               322.0        110.0                3.0  3.5  2.5  8.67   
6               330.0        115.0                5.0  4.5  3.0  9.34   

            research  chance of admit  
Serial No.                             
1                1.0             0.92  
2                1.0             0.76  
3                1.0             0.72  
4                1.0             0.80  
6                1.0             0.90

# Where and DropNa Together

In [29]:
df[df['chance of admit']>0.7].head()

gre score  toefl score  university rating  sop  lor  cgpa  \
Serial No.                                                              
1                 337          118                  4  4.5  4.5  9.65   
2                 324          107                  4  4.0  4.5  8.87   
3                 316          104                  3  3.0  3.5  8.00   
4                 322          110                  3  3.5  2.5  8.67   
6                 330          115                  5  4.5  3.0  9.34   

            research  chance of admit  
Serial No.                             
1                  1             0.92  
2                  1             0.76  
3                  1             0.72  
4                  1             0.80  
6                  1             0.90

In [30]:
df['chance of admit']

Serial No.
1      0.92
2      0.76
3      0.72
4      0.80
5      0.65
       ... 
396    0.82
397    0.84
398    0.91
399    0.67
400    0.95
Name: chance of admit, Length: 400, dtype: float64

In [31]:
df[['chance of admit', 'cgpa']].head(3)

chance of admit  cgpa
Serial No.                       
1                      0.92  9.65
2                      0.76  8.87
3                      0.72  8.00

note: each of them are mimicing functionality either .loc() or .where().dropna()

# Multiple Boolen and Multiple Criteria

#  AND `&`

In [8]:
df['chance of admit'].gt(0.7).lt(.9) #greater than and lesser than

Serial No.
1      False
2      False
3      False
4      False
5       True
       ...  
396    False
397    False
398    False
399     True
400    False
Name: chance of admit, Length: 400, dtype: bool

In [32]:
df[(df.cgpa>=7.00) & (df.cgpa<=8.87)].head(3)       

gre score  toefl score  university rating  sop  lor  cgpa  \
Serial No.                                                              
2                 324          107                  4  4.0  4.5  8.87   
3                 316          104                  3  3.0  3.5  8.00   
4                 322          110                  3  3.5  2.5  8.67   

            research  chance of admit  
Serial No.                             
2                  1             0.76  
3                  1             0.72  
4                  1             0.80

In [35]:
df.T[2]

gre score            324.00
toefl score          107.00
university rating      4.00
sop                    4.00
lor                    4.50
cgpa                   8.87
research               1.00
chance of admit        0.76
Name: 2, dtype: float64

In [45]:

#(df['cgpa'] > 7.00) & (df['cgpa'] < 8.87) df.where(df['toefl score'] > 105)
df.where(df['cgpa'] > 7.00).dropna()

gre score  toefl score  university rating  sop  lor  cgpa  \
Serial No.                                                              
1               337.0        118.0                4.0  4.5  4.5  9.65   
2               324.0        107.0                4.0  4.0  4.5  8.87   
3               316.0        104.0                3.0  3.0  3.5  8.00   
4               322.0        110.0                3.0  3.5  2.5  8.67   
5               314.0        103.0                2.0  2.0  3.0  8.21   
...               ...          ...                ...  ...  ...   ...   
396             324.0        110.0                3.0  3.5  3.5  9.04   
397             325.0        107.0                3.0  3.0  3.5  9.11   
398             330.0        116.0                4.0  5.0  4.5  9.45   
399             312.0        103.0                3.0  3.5  4.0  8.78   
400             333.0        117.0                4.0  5.0  4.0  9.66   

            research  chance of admit  
Serial No.                             
1                1.0             0.92  
2                1.0             0.76  
3                1.0             0.72  
4                1.0             0.80  
5                0.0             0.65  
...              ...              ...  
396              1.0             0.82  
397              1.0             0.84  
398              1.0             0.91  
399              0.0             0.67  
400              1.0             0.95  

[399 rows x 8 columns]

# set column as index


In [50]:
#preserve the original index as seial number
df['serial number'] = df.index

#now another column as index
df = df.set_index('chance of admit')
df.head(5)


gre score  toefl score  university rating  sop  lor  cgpa  \
chance of admit                                                              
0.92                   337          118                  4  4.5  4.5  9.65   
0.76                   324          107                  4  4.0  4.5  8.87   
0.72                   316          104                  3  3.0  3.5  8.00   
0.80                   322          110                  3  3.5  2.5  8.67   
0.65                   314          103                  2  2.0  3.0  8.21   

                 research  serial number  
chance of admit                           
0.92                    1              1  
0.76                    1              2  
0.72                    1              3  
0.80                    1              4  
0.65                    0              5

# reset_index

In [51]:
df=df.reset_index()
df.head(3)

chance of admit  gre score  toefl score  university rating  sop  lor  cgpa  \
0             0.92        337          118                  4  4.5  4.5  9.65   
1             0.76        324          107                  4  4.0  4.5  8.87   
2             0.72        316          104                  3  3.0  3.5  8.00   

   research  serial number  
0         1              1  
1         1              2  
2         1              3

note: here, chance 0f adimit promoted back to column after reseting index

# Demographics/Geological Data for Population

In [52]:
df = pd.read_csv ('resources/week-2/datasets/census.csv')
df.head(3)

SUMLEV  REGION  DIVISION  STATE  COUNTY   STNAME         CTYNAME  \
0      40       3         6      1       0  Alabama         Alabama   
1      50       3         6      1       1  Alabama  Autauga County   
2      50       3         6      1       3  Alabama  Baldwin County   

   CENSUS2010POP  ESTIMATESBASE2010  POPESTIMATE2010  ...  RDOMESTICMIG2011  \
0        4779736            4780127          4785161  ...          0.002295   
1          54571              54571            54660  ...          7.242091   
2         182265             182265           183193  ...         14.832960   

   RDOMESTICMIG2012  RDOMESTICMIG2013  RDOMESTICMIG2014  RDOMESTICMIG2015  \
0         -0.193196          0.381066          0.582002         -0.467369   
1         -2.915927         -3.012349          2.265971         -2.530799   
2         17.647293         21.845705         19.243287         17.197872   

   RNETMIG2011  RNETMIG2012  RNETMIG2013  RNETMIG2014  RNETMIG2015  
0     1.030015     0.826644     1.383282     1.724718     0.712594  
1     7.606016    -2.626146    -2.722002     2.592270    -2.187333  
2    15.844176    18.559627    22.727626    20.317142    18.293499  

[3 rows x 100 columns]

In [53]:
df.shape

(3193, 100)

In [54]:
df['SUMLEV'].unique()

array([40, 50])

In [55]:
#Now overwrite the data only for 50

df1 = df[df['SUMLEV']==50]
df1.head()

SUMLEV  REGION  DIVISION  STATE  COUNTY   STNAME         CTYNAME  \
1      50       3         6      1       1  Alabama  Autauga County   
2      50       3         6      1       3  Alabama  Baldwin County   
3      50       3         6      1       5  Alabama  Barbour County   
4      50       3         6      1       7  Alabama     Bibb County   
5      50       3         6      1       9  Alabama   Blount County   

   CENSUS2010POP  ESTIMATESBASE2010  POPESTIMATE2010  ...  RDOMESTICMIG2011  \
1          54571              54571            54660  ...          7.242091   
2         182265             182265           183193  ...         14.832960   
3          27457              27457            27341  ...         -4.728132   
4          22915              22919            22861  ...         -5.527043   
5          57322              57322            57373  ...          1.807375   

   RDOMESTICMIG2012  RDOMESTICMIG2013  RDOMESTICMIG2014  RDOMESTICMIG2015  \
1         -2.915927         -3.012349          2.265971         -2.530799   
2         17.647293         21.845705         19.243287         17.197872   
3         -2.500690         -7.056824         -3.904217        -10.543299   
4         -5.068871         -6.201001         -0.177537          0.177258   
5         -1.177622         -1.748766         -2.062535         -1.369970   

   RNETMIG2011  RNETMIG2012  RNETMIG2013  RNETMIG2014  RNETMIG2015  
1     7.606016    -2.626146    -2.722002     2.592270    -2.187333  
2    15.844176    18.559627    22.727626    20.317142    18.293499  
3    -4.874741    -2.758113    -7.167664    -3.978583   -10.543299  
4    -5.088389    -4.363636    -5.403729     0.754533     1.107861  
5     1.859511    -0.848580    -1.402476    -1.577232    -0.884411  

[5 rows x 100 columns]

Now reduce the dataframe and we only want to calculate the population based on birth til 2015

In [56]:
df1.columns


Index(['SUMLEV', 'REGION', 'DIVISION', 'STATE', 'COUNTY', 'STNAME', 'CTYNAME',
       'CENSUS2010POP', 'ESTIMATESBASE2010', 'POPESTIMATE2010',
       'POPESTIMATE2011', 'POPESTIMATE2012', 'POPESTIMATE2013',
       'POPESTIMATE2014', 'POPESTIMATE2015', 'NPOPCHG_2010', 'NPOPCHG_2011',
       'NPOPCHG_2012', 'NPOPCHG_2013', 'NPOPCHG_2014', 'NPOPCHG_2015',
       'BIRTHS2010', 'BIRTHS2011', 'BIRTHS2012', 'BIRTHS2013', 'BIRTHS2014',
       'BIRTHS2015', 'DEATHS2010', 'DEATHS2011', 'DEATHS2012', 'DEATHS2013',
       'DEATHS2014', 'DEATHS2015', 'NATURALINC2010', 'NATURALINC2011',
       'NATURALINC2012', 'NATURALINC2013', 'NATURALINC2014', 'NATURALINC2015',
       'INTERNATIONALMIG2010', 'INTERNATIONALMIG2011', 'INTERNATIONALMIG2012',
       'INTERNATIONALMIG2013', 'INTERNATIONALMIG2014', 'INTERNATIONALMIG2015',
       'DOMESTICMIG2010', 'DOMESTICMIG2011', 'DOMESTICMIG2012',
       'DOMESTICMIG2013', 'DOMESTICMIG2014', 'DOMESTICMIG2015', 'NETMIG2010',
       'NETMIG2011', 'NETMIG2012', 'NETMI

In [57]:
keeping_cols = ['STNAME', 'CTYNAME', 'BIRTHS2010', 'BIRTHS2011', 
               'BIRTHS2012', 'BIRTHS2013', 'BIRTHS2014',
               'BIRTHS2015', 'POPESTIMATE2010',
               'POPESTIMATE2011', 'POPESTIMATE2012', 'POPESTIMATE2013',
               'POPESTIMATE2014', 'POPESTIMATE2015' ]
df1 = df1[keeping_cols]
df1.head(3)

STNAME         CTYNAME  BIRTHS2010  BIRTHS2011  BIRTHS2012  BIRTHS2013  \
1  Alabama  Autauga County         151         636         615         574   
2  Alabama  Baldwin County         517        2187        2092        2160   
3  Alabama  Barbour County          70         335         300         283   

   BIRTHS2014  BIRTHS2015  POPESTIMATE2010  POPESTIMATE2011  POPESTIMATE2012  \
1         623         600            54660            55253            55175   
2        2186        2240           183193           186659           190396   
3         260         269            27341            27226            27159   

   POPESTIMATE2013  POPESTIMATE2014  POPESTIMATE2015  
1            55038            55290            55347  
2           195126           199713           203709  
3            26973            26815            26489

# Group Indexing

now set the index based on state and county/city value

In [58]:
#1st state name and then county/city name

df1 = df1.set_index(['STNAME', 'CTYNAME'])
df1.head()

BIRTHS2010  BIRTHS2011  BIRTHS2012  BIRTHS2013  \
STNAME  CTYNAME                                                          
Alabama Autauga County         151         636         615         574   
        Baldwin County         517        2187        2092        2160   
        Barbour County          70         335         300         283   
        Bibb County             44         266         245         259   
        Blount County          183         744         710         646   

                        BIRTHS2014  BIRTHS2015  POPESTIMATE2010  \
STNAME  CTYNAME                                                   
Alabama Autauga County         623         600            54660   
        Baldwin County        2186        2240           183193   
        Barbour County         260         269            27341   
        Bibb County            247         253            22861   
        Blount County          618         603            57373   

                        POPESTIMATE2011  POPESTIMATE2012  POPESTIMATE2013  \
STNAME  CTYNAME                                                             
Alabama Autauga County            55253            55175            55038   
        Baldwin County           186659           190396           195126   
        Barbour County            27226            27159            26973   
        Bibb County               22733            22642            22512   
        Blount County             57711            57776            57734   

                        POPESTIMATE2014  POPESTIMATE2015  
STNAME  CTYNAME                                           
Alabama Autauga County            55290            55347  
        Baldwin County           199713           203709  
        Barbour County            26815            26489  
        Bibb County               22549            22583  
        Blount County             57658            57673

In [59]:
df1.index

MultiIndex([('Alabama',    'Autauga County'),
            ('Alabama',    'Baldwin County'),
            ('Alabama',    'Barbour County'),
            ('Alabama',       'Bibb County'),
            ('Alabama',     'Blount County'),
            ('Alabama',    'Bullock County'),
            ('Alabama',     'Butler County'),
            ('Alabama',    'Calhoun County'),
            ('Alabama',   'Chambers County'),
            ('Alabama',   'Cherokee County'),
            ...
            ('Wyoming',   'Niobrara County'),
            ('Wyoming',       'Park County'),
            ('Wyoming',     'Platte County'),
            ('Wyoming',   'Sheridan County'),
            ('Wyoming',   'Sublette County'),
            ('Wyoming', 'Sweetwater County'),
            ('Wyoming',      'Teton County'),
            ('Wyoming',      'Uinta County'),
            ('Wyoming',   'Washakie County'),
            ('Wyoming',     'Weston County')],
           names=['STNAME', 'CTYNAME'], length=3142)

In [60]:
# we need to call mutiple index to get any certain result by maintaining the chronological order

df1.loc['Alabama', 'Chambers County']

BIRTHS2010            81
BIRTHS2011           401
BIRTHS2012           393
BIRTHS2013           404
BIRTHS2014           421
BIRTHS2015           429
POPESTIMATE2010    34098
POPESTIMATE2011    33993
POPESTIMATE2012    34075
POPESTIMATE2013    34153
POPESTIMATE2014    34052
POPESTIMATE2015    34123
Name: (Alabama, Chambers County), dtype: int64

# Find two county/ctyname against a state 

In [62]:
df1.loc[[('Alabama', 'Chambers County'), ('Wyoming', 'Niobrara County')]]

BIRTHS2010  BIRTHS2011  BIRTHS2012  BIRTHS2013  \
STNAME  CTYNAME                                                           
Alabama Chambers County          81         401         393         404   
Wyoming Niobrara County           6          18          18          27   

                         BIRTHS2014  BIRTHS2015  POPESTIMATE2010  \
STNAME  CTYNAME                                                    
Alabama Chambers County         421         429            34098   
Wyoming Niobrara County          27          25             2492   

                         POPESTIMATE2011  POPESTIMATE2012  POPESTIMATE2013  \
STNAME  CTYNAME                                                              
Alabama Chambers County            33993            34075            34153   
Wyoming Niobrara County             2485             2475             2548   

                         POPESTIMATE2014  POPESTIMATE2015  
STNAME  CTYNAME                                            
Alabama Chambers County            34052            34123  
Wyoming Niobrara County             2530             2542

# Missing Value

**Missing at random** missing value relates to other columns/variables

**Missing Completely at random** missing value is completely independent from other columns/variables

Pandas handles missing data `None type` and `Numpy NaN values`

`Format` of missing values : NaN, NULL, None or N/A
in social scientists, who are regularly using value of **`99`**

In [1]:
import pandas as pd

In [2]:

df =pd.read_csv('resources/week-2/datasets/class_grades.csv')
df.head(10)

Prefix  Assignment  Tutorial  Midterm  TakeHome  Final
0       5       57.14     34.09    64.38     51.48  52.50
1       8       95.05    105.49    67.50     99.07  68.33
2       8       83.70     83.17      NaN     63.15  48.89
3       7         NaN       NaN    49.38    105.93  80.56
4       8       91.32     93.64    95.00    107.41  73.89
5       7       95.00     92.58    93.12     97.78  68.06
6       8       95.05    102.99    56.25     99.07  50.00
7       7       72.85     86.85    60.00       NaN  56.11
8       8       84.26     93.10    47.50     18.52  50.83
9       7       90.10     97.55    51.25     88.89  63.61

# **`.isnull()`**

In [3]:
mask = df.isnull()
mask.head(10)

Prefix  Assignment  Tutorial  Midterm  TakeHome  Final
0   False       False     False    False     False  False
1   False       False     False    False     False  False
2   False       False     False     True     False  False
3   False        True      True    False     False  False
4   False       False     False    False     False  False
5   False       False     False    False     False  False
6   False       False     False    False     False  False
7   False       False     False    False      True  False
8   False       False     False    False     False  False
9   False       False     False    False     False  False

# **`.dropna()`**

In [5]:
df.dropna().head(10)

Prefix  Assignment  Tutorial  Midterm  TakeHome  Final
0        5       57.14     34.09    64.38     51.48  52.50
1        8       95.05    105.49    67.50     99.07  68.33
4        8       91.32     93.64    95.00    107.41  73.89
5        7       95.00     92.58    93.12     97.78  68.06
6        8       95.05    102.99    56.25     99.07  50.00
8        8       84.26     93.10    47.50     18.52  50.83
9        7       90.10     97.55    51.25     88.89  63.61
10       7       80.44     90.20    75.00     91.48  39.72
12       8       97.16    103.71    72.50     93.52  63.33
13       7       91.28     83.53    81.25     99.81  92.22

# **`.fillna(0, inplace=True)`**

0 is a value to replace the NaN value

In [6]:
df.fillna(0, inplace = True)

In [6]:
df.head(10)

Prefix  Assignment  Tutorial  Midterm  TakeHome  Final
0       5       57.14     34.09    64.38     51.48  52.50
1       8       95.05    105.49    67.50     99.07  68.33
2       8       83.70     83.17     0.00     63.15  48.89
3       7        0.00      0.00    49.38    105.93  80.56
4       8       91.32     93.64    95.00    107.41  73.89
5       7       95.00     92.58    93.12     97.78  68.06
6       8       95.05    102.99    56.25     99.07  50.00
7       7       72.85     86.85    60.00      0.00  56.11
8       8       84.26     93.10    47.50     18.52  50.83
9       7       90.10     97.55    51.25     88.89  63.61

# Log DataFrame

In [15]:
df = pd.read_csv("resources/week-2/datasets/log.csv")
df.head(10)

time    user       video  playback position paused  volume
0  1469974424  cheryl  intro.html                  5  False    10.0
1  1469974454  cheryl  intro.html                  6    NaN     NaN
2  1469974544  cheryl  intro.html                  9    NaN     NaN
3  1469974574  cheryl  intro.html                 10    NaN     NaN
4  1469977514     bob  intro.html                  1    NaN     NaN
5  1469977544     bob  intro.html                  1    NaN     NaN
6  1469977574     bob  intro.html                  1    NaN     NaN
7  1469977604     bob  intro.html                  1    NaN     NaN
8  1469974604  cheryl  intro.html                 11    NaN     NaN
9  1469974694  cheryl  intro.html                 14    NaN     NaN

# **`ffill()`** and **`bfill()`**

**`ffill()`** :: **Forward fill** na value replaces by a previous value from a row  of that particular column.

**`bfill()`** :: **Backward fill** na value replaces by a next valid value from a row of that particular column.

# Indexing and Sorting

In [16]:
#setting index
df = df.set_index('time')

#sorting index by time
df = df.sort_index()


df.head()

user          video  playback position paused  volume
time                                                               
1469974424  cheryl     intro.html                  5  False    10.0
1469974424     sue  advanced.html                 23  False    10.0
1469974454  cheryl     intro.html                  6    NaN     NaN
1469974454     sue  advanced.html                 24    NaN     NaN
1469974484  cheryl     intro.html                  7    NaN     NaN

In [17]:
df.reset_index(inplace=True)
df.head(3)

time    user          video  playback position paused  volume
0  1469974424  cheryl     intro.html                  5  False    10.0
1  1469974424     sue  advanced.html                 23  False    10.0
2  1469974454  cheryl     intro.html                  6    NaN     NaN

In [18]:
df.columns

Index(['time', 'user', 'video', 'playback position', 'paused', 'volume'], dtype='object')

In [19]:
df.set_index(['time', 'user']).head(5)

video  playback position paused  volume
time       user                                                   
1469974424 cheryl     intro.html                  5  False    10.0
           sue     advanced.html                 23  False    10.0
1469974454 cheryl     intro.html                  6    NaN     NaN
           sue     advanced.html                 24    NaN     NaN
1469974484 cheryl     intro.html                  7    NaN     NaN

In [20]:
df = df.fillna(method='ffill')
df.head()

time    user          video  playback position  paused  volume
0  1469974424  cheryl     intro.html                  5   False    10.0
1  1469974424     sue  advanced.html                 23   False    10.0
2  1469974454  cheryl     intro.html                  6   False    10.0
3  1469974454     sue  advanced.html                 24   False    10.0
4  1469974484  cheryl     intro.html                  7   False    10.0

# Replace

In [30]:
store = pd.DataFrame({ 'A': [1,2,3,4,5],
                      'B': [1,12,13,14,15],
                      'C': ['a', 'b', 'c', 'd', 'e']}
)
store.head(5)

A   B  C
0  1   1  a
1  2  12  b
2  3  13  c
3  4  14  d
4  5  15  e

Replace value **1** with **100**

In [31]:
store.replace(1, 100)

A    B  C
0  100  100  a
1    2   12  b
2    3   13  c
3    4   14  d
4    5   15  e

# Replace multiple value at a time

**`.replace([existing value 1, existing value 2], [new value 1, new value 2])`**

In [33]:
store.replace([2, 14], [70, 350])

A    B  C
0   1    1  a
1  70   12  b
2   3   13  c
3   4  350  d
4   5   15  e

# Replace using Regex

In [40]:
import pandas as pd

In [44]:
df = pd.read_csv("resources/week-2/datasets/log.csv")
df.head(10)


time    user       video  playback position paused  volume
0  1469974424  cheryl  intro.html                  5  False    10.0
1  1469974454  cheryl  intro.html                  6    NaN     NaN
2  1469974544  cheryl  intro.html                  9    NaN     NaN
3  1469974574  cheryl  intro.html                 10    NaN     NaN
4  1469977514     bob  intro.html                  1    NaN     NaN
5  1469977544     bob  intro.html                  1    NaN     NaN
6  1469977574     bob  intro.html                  1    NaN     NaN
7  1469977604     bob  intro.html                  1    NaN     NaN
8  1469974604  cheryl  intro.html                 11    NaN     NaN
9  1469974694  cheryl  intro.html                 14    NaN     NaN

`$` refers at the end of each value consisting **.html**

In [46]:
df.replace(to_replace= " *.html$", value = ' webpage', regex=True)

time    user             video  playback position paused  volume
0   1469974424  cheryl     intro webpage                  5  False    10.0
1   1469974454  cheryl     intro webpage                  6    NaN     NaN
2   1469974544  cheryl     intro webpage                  9    NaN     NaN
3   1469974574  cheryl     intro webpage                 10    NaN     NaN
4   1469977514     bob     intro webpage                  1    NaN     NaN
5   1469977544     bob     intro webpage                  1    NaN     NaN
6   1469977574     bob     intro webpage                  1    NaN     NaN
7   1469977604     bob     intro webpage                  1    NaN     NaN
8   1469974604  cheryl     intro webpage                 11    NaN     NaN
9   1469974694  cheryl     intro webpage                 14    NaN     NaN
10  1469974724  cheryl     intro webpage                 15    NaN     NaN
11  1469974454     sue  advanced webpage                 24    NaN     NaN
12  1469974524     sue  advanced webpage                 25    NaN     NaN
13  1469974424     sue  advanced webpage                 23  False    10.0
14  1469974554     sue  advanced webpage                 26    NaN     NaN
15  1469974624     sue  advanced webpage                 27    NaN     NaN
16  1469974654     sue  advanced webpage                 28    NaN     5.0
17  1469974724     sue  advanced webpage                 29    NaN     NaN
18  1469974484  cheryl     intro webpage                  7    NaN     NaN
19  1469974514  cheryl     intro webpage                  8    NaN     NaN
20  1469974754     sue  advanced webpage                 30    NaN     NaN
21  1469974824     sue  advanced webpage                 31    NaN     NaN
22  1469974854     sue  advanced webpage                 32    NaN     NaN
23  1469974924     sue  advanced webpage                 33    NaN     NaN
24  1469977424     bob     intro webpage                  1   True    10.0
25  1469977454     bob     intro webpage                  1    NaN     NaN
26  1469977484     bob     intro webpage                  1    NaN     NaN
27  1469977634     bob     intro webpage                  1    NaN     NaN
28  1469977664     bob     intro webpage                  1    NaN     NaN
29  1469974634  cheryl     intro webpage                 12    NaN     NaN
30  1469974664  cheryl     intro webpage                 13    NaN     NaN
31  1469977694     bob     intro webpage                  1    NaN     NaN
32  1469977724     bob     intro webpage                  1    NaN     NaN

# Manipulating DataFrame

In [2]:
import pandas as pd

In [3]:
df = pd.read_csv("resources/week-2/datasets/presidents.csv")
df.head(5)

#          President             Born      Age atstart of presidency  \
0  1  George Washington  Feb 22, 1732[a]  57 years, 67 daysApr 30, 1789   
1  2         John Adams  Oct 30, 1735[a]  61 years, 125 daysMar 4, 1797   
2  3   Thomas Jefferson  Apr 13, 1743[a]  57 years, 325 daysMar 4, 1801   
3  4      James Madison  Mar 16, 1751[a]  57 years, 353 daysMar 4, 1809   
4  5       James Monroe     Apr 28, 1758  58 years, 310 daysMar 4, 1817   

         Age atend of presidency Post-presidencytimespan          Died  \
0   65 years, 10 daysMar 4, 1797       2 years, 285 days  Dec 14, 1799   
1  65 years, 125 daysMar 4, 1801      25 years, 122 days   Jul 4, 1826   
2  65 years, 325 daysMar 4, 1809      17 years, 122 days   Jul 4, 1826   
3  65 years, 353 daysMar 4, 1817      19 years, 116 days  Jun 28, 1836   
4  66 years, 310 daysMar 4, 1825       6 years, 122 days   Jul 4, 1831   

                  Age  
0  67 years, 295 days  
1  90 years, 247 days  
2   83 years, 82 days  
3  85 years, 104 days  
4   73 years, 67 days

`*` refers for selecting the **First Name** from Full Name

In [4]:
df['First'] = df['President']

In [5]:
df['First'] = df['First'].replace("[ ].*", "",regex=True)
df.head(5)

#          President             Born      Age atstart of presidency  \
0  1  George Washington  Feb 22, 1732[a]  57 years, 67 daysApr 30, 1789   
1  2         John Adams  Oct 30, 1735[a]  61 years, 125 daysMar 4, 1797   
2  3   Thomas Jefferson  Apr 13, 1743[a]  57 years, 325 daysMar 4, 1801   
3  4      James Madison  Mar 16, 1751[a]  57 years, 353 daysMar 4, 1809   
4  5       James Monroe     Apr 28, 1758  58 years, 310 daysMar 4, 1817   

         Age atend of presidency Post-presidencytimespan          Died  \
0   65 years, 10 daysMar 4, 1797       2 years, 285 days  Dec 14, 1799   
1  65 years, 125 daysMar 4, 1801      25 years, 122 days   Jul 4, 1826   
2  65 years, 325 daysMar 4, 1809      17 years, 122 days   Jul 4, 1826   
3  65 years, 353 daysMar 4, 1817      19 years, 116 days  Jun 28, 1836   
4  66 years, 310 daysMar 4, 1825       6 years, 122 days   Jul 4, 1831   

                  Age   First  
0  67 years, 295 days  George  
1  90 years, 247 days    John  
2   83 years, 82 days  Thomas  
3  85 years, 104 days   James  
4   73 years, 67 days   James

# **`del`**

In [6]:
del(df['First'])
df.head(3)

#          President             Born      Age atstart of presidency  \
0  1  George Washington  Feb 22, 1732[a]  57 years, 67 daysApr 30, 1789   
1  2         John Adams  Oct 30, 1735[a]  61 years, 125 daysMar 4, 1797   
2  3   Thomas Jefferson  Apr 13, 1743[a]  57 years, 325 daysMar 4, 1801   

         Age atend of presidency Post-presidencytimespan          Died  \
0   65 years, 10 daysMar 4, 1797       2 years, 285 days  Dec 14, 1799   
1  65 years, 125 daysMar 4, 1801      25 years, 122 days   Jul 4, 1826   
2  65 years, 325 daysMar 4, 1809      17 years, 122 days   Jul 4, 1826   

                  Age  
0  67 years, 295 days  
1  90 years, 247 days  
2   83 years, 82 days

# **`.apply()`**

In [7]:
def splitName(row):
    row['First'] = row['President'].split(" ")[0]
    row['Last'] = row['President'].split(" ")[-1]
    
    return row
df=df.apply(splitName, axis = 'columns')
df.head(3)

#          President             Born      Age atstart of presidency  \
0  1  George Washington  Feb 22, 1732[a]  57 years, 67 daysApr 30, 1789   
1  2         John Adams  Oct 30, 1735[a]  61 years, 125 daysMar 4, 1797   
2  3   Thomas Jefferson  Apr 13, 1743[a]  57 years, 325 daysMar 4, 1801   

         Age atend of presidency Post-presidencytimespan          Died  \
0   65 years, 10 daysMar 4, 1797       2 years, 285 days  Dec 14, 1799   
1  65 years, 125 daysMar 4, 1801      25 years, 122 days   Jul 4, 1826   
2  65 years, 325 daysMar 4, 1809      17 years, 122 days   Jul 4, 1826   

                  Age   First        Last  
0  67 years, 295 days  George  Washington  
1  90 years, 247 days    John       Adams  
2   83 years, 82 days  Thomas   Jefferson

In [8]:
del(df['First'])
del(df['Last'])
df.head(3)

#          President             Born      Age atstart of presidency  \
0  1  George Washington  Feb 22, 1732[a]  57 years, 67 daysApr 30, 1789   
1  2         John Adams  Oct 30, 1735[a]  61 years, 125 daysMar 4, 1797   
2  3   Thomas Jefferson  Apr 13, 1743[a]  57 years, 325 daysMar 4, 1801   

         Age atend of presidency Post-presidencytimespan          Died  \
0   65 years, 10 daysMar 4, 1797       2 years, 285 days  Dec 14, 1799   
1  65 years, 125 daysMar 4, 1801      25 years, 122 days   Jul 4, 1826   
2  65 years, 325 daysMar 4, 1809      17 years, 122 days   Jul 4, 1826   

                  Age  
0  67 years, 295 days  
1  90 years, 247 days  
2   83 years, 82 days

# **`.extract()`**

there are 3 groups to retrive only two groups

`([\w]*)` : First Name

`(?:.*)` : **`?:`** is used not to retrun the value **`.*`** any type of whitespace

`([\w]*$)`: Last Name

In [9]:
df.head(3)

#          President             Born      Age atstart of presidency  \
0  1  George Washington  Feb 22, 1732[a]  57 years, 67 daysApr 30, 1789   
1  2         John Adams  Oct 30, 1735[a]  61 years, 125 daysMar 4, 1797   
2  3   Thomas Jefferson  Apr 13, 1743[a]  57 years, 325 daysMar 4, 1801   

         Age atend of presidency Post-presidencytimespan          Died  \
0   65 years, 10 daysMar 4, 1797       2 years, 285 days  Dec 14, 1799   
1  65 years, 125 daysMar 4, 1801      25 years, 122 days   Jul 4, 1826   
2  65 years, 325 daysMar 4, 1809      17 years, 122 days   Jul 4, 1826   

                  Age  
0  67 years, 295 days  
1  90 years, 247 days  
2   83 years, 82 days

In [10]:
pattern = "(^[\w]*)(?:.* )([\w]*$)"
df['President'].str.extract(pattern).head()

0           1
0  George  Washington
1    John       Adams
2  Thomas   Jefferson
3   James     Madison
4   James      Monroe

# Identify pattern with column Name

**`"(?P<First>^[\w]*)(?:.* )(?P<Last>[\w]*$)"`**

In [11]:
pattern = "(?P<First>^[\w]*)(?:.* )(?P<Last>[\w]*$)"
name = df['President'].str.extract(pattern).head()
name.head(3)

First        Last
0  George  Washington
1    John       Adams
2  Thomas   Jefferson

In [12]:
df['First'] = name['First']
df['Last'] = name['Last']

In [13]:
df.head(3)

#          President             Born      Age atstart of presidency  \
0  1  George Washington  Feb 22, 1732[a]  57 years, 67 daysApr 30, 1789   
1  2         John Adams  Oct 30, 1735[a]  61 years, 125 daysMar 4, 1797   
2  3   Thomas Jefferson  Apr 13, 1743[a]  57 years, 325 daysMar 4, 1801   

         Age atend of presidency Post-presidencytimespan          Died  \
0   65 years, 10 daysMar 4, 1797       2 years, 285 days  Dec 14, 1799   
1  65 years, 125 daysMar 4, 1801      25 years, 122 days   Jul 4, 1826   
2  65 years, 325 daysMar 4, 1809      17 years, 122 days   Jul 4, 1826   

                  Age   First        Last  
0  67 years, 295 days  George  Washington  
1  90 years, 247 days    John       Adams  
2   83 years, 82 days  Thomas   Jefferson

**`""([\w]{3} [\w]{1,2}, [\w]{4})""`** 

[\w]{3} for Month **Feb** followed by space then [\w]{1,2} for date **22** use **`,`** then space followd by [\w]{4} for year **1732**

In [15]:
df['Born'] = df['Born'].str.extract("([\w]{3} [\w]{1,2}, [\w]{4})")

In [16]:
df.head(3)

#          President          Born      Age atstart of presidency  \
0  1  George Washington  Feb 22, 1732  57 years, 67 daysApr 30, 1789   
1  2         John Adams  Oct 30, 1735  61 years, 125 daysMar 4, 1797   
2  3   Thomas Jefferson  Apr 13, 1743  57 years, 325 daysMar 4, 1801   

         Age atend of presidency Post-presidencytimespan          Died  \
0   65 years, 10 daysMar 4, 1797       2 years, 285 days  Dec 14, 1799   
1  65 years, 125 daysMar 4, 1801      25 years, 122 days   Jul 4, 1826   
2  65 years, 325 daysMar 4, 1809      17 years, 122 days   Jul 4, 1826   

                  Age   First        Last  
0  67 years, 295 days  George  Washington  
1  90 years, 247 days    John       Adams  
2   83 years, 82 days  Thomas   Jefferson

# Pandas DataTime

In [17]:
df['Born'] = pd.to_datetime(df['Born'])
df['Born'].head(4)

0   1732-02-22
1   1735-10-30
2   1743-04-13
3   1751-03-16
Name: Born, dtype: datetime64[ns]

In [18]:
df.head(3)

#          President       Born      Age atstart of presidency  \
0  1  George Washington 1732-02-22  57 years, 67 daysApr 30, 1789   
1  2         John Adams 1735-10-30  61 years, 125 daysMar 4, 1797   
2  3   Thomas Jefferson 1743-04-13  57 years, 325 daysMar 4, 1801   

         Age atend of presidency Post-presidencytimespan          Died  \
0   65 years, 10 daysMar 4, 1797       2 years, 285 days  Dec 14, 1799   
1  65 years, 125 daysMar 4, 1801      25 years, 122 days   Jul 4, 1826   
2  65 years, 325 daysMar 4, 1809      17 years, 122 days   Jul 4, 1826   

                  Age   First        Last  
0  67 years, 295 days  George  Washington  
1  90 years, 247 days    John       Adams  
2   83 years, 82 days  Thomas   Jefferson

In [32]:
df['Died'].unique

<bound method Series.unique of 0     Dec 14, 1799
1      Jul 4, 1826
2      Jul 4, 1826
3     Jun 28, 1836
4      Jul 4, 1831
5     Feb 23, 1848
6      Jun 8, 1845
7     Jul 24, 1862
8      Apr 4, 1841
9     Jan 18, 1862
10    Jun 15, 1849
11     Jul 9, 1850
12     Mar 8, 1874
13     Oct 8, 1869
14     Jun 1, 1868
15    Apr 15, 1865
16    Jul 31, 1875
17    Jul 23, 1885
18    Jan 17, 1893
19    Sep 19, 1881
20    Nov 18, 1886
21    Jun 24, 1908
22    Mar 13, 1901
23    Jun 24, 1908
24    Sep 14, 1901
25     Jan 6, 1919
26     Mar 8, 1930
27     Feb 3, 1924
28     Aug 2, 1923
29     Jan 5, 1933
30    Oct 20, 1964
31    Apr 12, 1945
32    Dec 26, 1972
33    Mar 28, 1969
34    Nov 22, 1963
35    Jan 22, 1973
36    Apr 22, 1994
37    Dec 26, 2006
38        (living)
39     Jun 5, 2004
40    Nov 30, 2018
41        (living)
42        (living)
43        (living)
Name: Died, dtype: object>

In [34]:
df[df['Died']=='(living)']

#       President       Born       Age atstart of presidency  \
38  39    Jimmy Carter 1924-10-01  52 years, 111 daysJan 20, 1977   
41  42    Bill Clinton 1946-08-19  46 years, 154 daysJan 20, 1993   
42  43  George W. Bush 1946-07-06  54 years, 198 daysJan 20, 2001   
43  44    Barack Obama 1961-08-04  47 years, 169 daysJan 20, 2009   

           Age atend of presidency Post-presidencytimespan      Died  \
38  56 years, 111 daysJan 20, 1981      38 years, 175 days  (living)   
41  54 years, 154 daysJan 20, 2001      18 years, 175 days  (living)   
42  62 years, 198 daysJan 20, 2009      10 years, 175 days  (living)   
43  55 years, 169 daysJan 20, 2017       2 years, 175 days  (living)   

                   Age First Last  
38  94 years, 286 days   NaN  NaN  
41  72 years, 329 days   NaN  NaN  
42    73 years, 8 days   NaN  NaN  
43  57 years, 344 days   NaN  NaN

# Practice

In [40]:
import pandas as pd
import pandas as pd
sdata = {'Ohio': 35000, 'Texas': 71000, 'Oregon': 16000, 'Utah': 5000}
obj1 = pd.Series(sdata)
states = ['California', 'Ohio', 'Oregon', 'Texas']
obj2 = pd.Series(sdata, index=states)
obj3 = pd.isnull(obj2)

In [41]:
obj2['California'] == None

False

In [16]:
import pandas as pd
s1 = pd.Series({1: 'Alice', 2: 'Jack', 3: 'Molly'})
s2 = pd.Series({'Alice': 1, 'Jack': 2, 'Molly': 3})

In [19]:
s2.iloc[1]

2

In [ ]:
s2[1]

In [15]:
l = pd.DataFrame([['A', 'B', 'C'], [11,12,13]])
l

0   1   2
0   A   B   C
1  11  12  13

In [14]:
l.drop(1)

0  1  2
0  A  B  C

In [38]:
import pandas as pd
d = {'1': 'Alice','2': 'Bob','3': 'Rita','4': 'Molly','5': 'Ryan'}
S = pd.Series(d)

In [39]:

S.iloc[0:3]

1    Alice
2      Bob
3     Rita
dtype: object

In [36]:
m=pd.Series([1,2,4,5])
n=pd.Series([11,21,41,51])

In [37]:
pd.DataFrame(m)

0
0  1
1  2
2  4
3  5

In [23]:
n.append(m)

0    11
1    21
2    41
3    51
0     1
1     2
2     4
3     5
dtype: int64

In [46]:
import pandas as pd
s1 = pd.Series({1: 'Alice', 2: 'Jack', 3: 'Molly'})
s2 = pd.Series({'Alice': 1, 'Jack': 2, 'Molly': 3})

In [47]:

s1.loc[1]

'Alice'

# Assignment -2

# A:

Write a function called proportion_of_education which returns the proportion of children in the dataset who had a mother with the education levels equal to less than high school (<12), high school (12), more than high school but not a college graduate (>12) and college degree.

This function should return a dictionary in the form of (use the correct numbers, do not round numbers):

    {"less than high school":0.2,
   
     "high school":0.4,
    
     "more than high school but not college":0.2,
    
     "college":0.2}

In [2]:
import pandas as pd
import numpy as np

In [3]:
df = pd.read_csv("assignments/assignment2/assets/NISPUF17.csv")
df.head(4)

Unnamed: 0  SEQNUMC  SEQNUMHH  PDAT    PROVWT_D     RDDWT_D  STRATUM  YEAR  \
0           1   128521     12852     2         NaN  235.916956     1031  2017   
1           2    10741      1074     2         NaN  957.353840     1068  2017   
2           3   220011     22001     2         NaN  189.611299     1050  2017   
3           4    86131      8613     1  675.430817  333.447418     1040  2017   

   AGECPOXR  HAD_CPOX  ...  XVRCTY2  XVRCTY3  XVRCTY4  XVRCTY5  XVRCTY6  \
0       NaN         2  ...                        NaN      NaN      NaN   
1       NaN         2  ...                        NaN      NaN      NaN   
2       NaN         2  ...                        NaN      NaN      NaN   
3       NaN         2  ...                        NaN      NaN      NaN   

   XVRCTY7  XVRCTY8  XVRCTY9  INS_STAT2_I  INS_BREAK_I  
0      NaN      NaN      NaN          NaN          NaN  
1      NaN      NaN      NaN          NaN          NaN  
2      NaN      NaN      NaN          NaN          NaN  
3      NaN      NaN      NaN          1.0          2.0  

[4 rows x 454 columns]

In [4]:
df.columns

Index(['Unnamed: 0', 'SEQNUMC', 'SEQNUMHH', 'PDAT', 'PROVWT_D', 'RDDWT_D',
       'STRATUM', 'YEAR', 'AGECPOXR', 'HAD_CPOX',
       ...
       'XVRCTY2', 'XVRCTY3', 'XVRCTY4', 'XVRCTY5', 'XVRCTY6', 'XVRCTY7',
       'XVRCTY8', 'XVRCTY9', 'INS_STAT2_I', 'INS_BREAK_I'],
      dtype='object', length=454)

In [5]:
df['EDUC1']

0        4
1        3
2        3
3        4
4        1
        ..
28460    3
28461    2
28462    3
28463    2
28464    4
Name: EDUC1, Length: 28465, dtype: int64

In [6]:
df = pd.read_csv("assignments/assignment2/assets/NISPUF17.csv")


In [7]:
df['EDUC1']

0        4
1        3
2        3
3        4
4        1
        ..
28460    3
28461    2
28462    3
28463    2
28464    4
Name: EDUC1, Length: 28465, dtype: int64

In [8]:
df['EDUC1'].unique()

array([4, 3, 1, 2])

In [9]:
len(df['EDUC1'])

28465

In [10]:
dict={"less than high school":0,
    "high school":0,
    "more than high school but not college":0,
    "college":0}
    
total = len(df['EDUC1'])
educ = df['EDUC1'].value_counts().sort_index().to_list()


dict["less than high school"] = educ[0]/total
dict["high school"] = educ[1]/total
dict["more than high school but not college"] = educ[2]/total
dict["college"] = educ[3]/total

In [11]:
dict

{'less than high school': 0.10202002459160373,
 'high school': 0.172352011241876,
 'more than high school but not college': 0.24588090637625154,
 'college': 0.47974705779026877}

# B:

Let's explore the relationship between being fed breastmilk as a child and getting a seasonal influenza vaccine from a healthcare provider. Return a tuple of the average number of influenza vaccines for those children we know received breastmilk as a child and those who know did not.

This function should return a tuple in the form (use the correct numbers:

(2.5, 0.1)

In [2]:
import pandas as pd
import numpy as np

df = pd.read_csv("assignments/assignment2/assets/NISPUF17.csv")
df.head(4)


Unnamed: 0  SEQNUMC  SEQNUMHH  PDAT    PROVWT_D     RDDWT_D  STRATUM  YEAR  \
0           1   128521     12852     2         NaN  235.916956     1031  2017   
1           2    10741      1074     2         NaN  957.353840     1068  2017   
2           3   220011     22001     2         NaN  189.611299     1050  2017   
3           4    86131      8613     1  675.430817  333.447418     1040  2017   

   AGECPOXR  HAD_CPOX  ...  XVRCTY2  XVRCTY3  XVRCTY4  XVRCTY5  XVRCTY6  \
0       NaN         2  ...                        NaN      NaN      NaN   
1       NaN         2  ...                        NaN      NaN      NaN   
2       NaN         2  ...                        NaN      NaN      NaN   
3       NaN         2  ...                        NaN      NaN      NaN   

   XVRCTY7  XVRCTY8  XVRCTY9  INS_STAT2_I  INS_BREAK_I  
0      NaN      NaN      NaN          NaN          NaN  
1      NaN      NaN      NaN          NaN          NaN  
2      NaN      NaN      NaN          NaN          NaN  
3      NaN      NaN      NaN          1.0          2.0  

[4 rows x 454 columns]

In [7]:
df['CBF_01'].value_counts()

1     24261
2      4085
77      110
99        9
Name: CBF_01, dtype: int64

In [10]:
df['P_NUMFLU'].value_counts()

3.0    4845
0.0    4139
2.0    3350
1.0    1679
4.0    1282
5.0      35
6.0       3
Name: P_NUMFLU, dtype: int64

In [14]:
influ_df = df[['CBF_01', 'P_NUMFLU']]
influ_df.head(3)

CBF_01  P_NUMFLU
0       1       NaN
1       2       NaN
2       2       NaN

In [20]:
influ_df = influ_df.dropna()

influ_df.head(3)

CBF_01  P_NUMFLU
3       2       3.0
4       1       0.0
6       1       0.0

In [26]:
cf1 = influ_df[influ_df['CBF_01'] == 1]
f1 = cf1['P_NUMFLU'].mean()
f1

1.8799187420058687

In [29]:
cf2 = influ_df[influ_df['CBF_01']  == 2]
f2 = cf2['P_NUMFLU'].mean()
f2

1.5963945918878317

In [32]:
print(f1,f2)

1.8799187420058687 1.5963945918878317


# C: 

It would be interesting to see if there is any evidence of a link between vaccine effectiveness and sex of the child. Calculate the ratio of the number of children who contracted chickenpox but were vaccinated against it (at least one varicella dose) versus those who were vaccinated but did not contract chicken pox. Return results by sex.

This function should return a dictionary in the form of (use the correct numbers):

    {"male":0.2,
    "female":0.4}


In [12]:
import pandas as pd
import numpy as np

df = pd.read_csv("assignments/assignment2/assets/NISPUF17.csv")
df.head(4)

Unnamed: 0  SEQNUMC  SEQNUMHH  PDAT    PROVWT_D     RDDWT_D  STRATUM  YEAR  \
0           1   128521     12852     2         NaN  235.916956     1031  2017   
1           2    10741      1074     2         NaN  957.353840     1068  2017   
2           3   220011     22001     2         NaN  189.611299     1050  2017   
3           4    86131      8613     1  675.430817  333.447418     1040  2017   

   AGECPOXR  HAD_CPOX  ...  XVRCTY2  XVRCTY3  XVRCTY4  XVRCTY5  XVRCTY6  \
0       NaN         2  ...                        NaN      NaN      NaN   
1       NaN         2  ...                        NaN      NaN      NaN   
2       NaN         2  ...                        NaN      NaN      NaN   
3       NaN         2  ...                        NaN      NaN      NaN   

   XVRCTY7  XVRCTY8  XVRCTY9  INS_STAT2_I  INS_BREAK_I  
0      NaN      NaN      NaN          NaN          NaN  
1      NaN      NaN      NaN          NaN          NaN  
2      NaN      NaN      NaN          NaN          NaN  
3      NaN      NaN      NaN          1.0          2.0  

[4 rows x 454 columns]

In [36]:
df['HAD_CPOX'].value_counts()

2     27955
1       402
77      105
99        3
Name: HAD_CPOX, dtype: int64

In [40]:
df['SEX'].value_counts()

1    14717
2    13748
Name: SEX, dtype: int64

In [42]:
df['P_NUMVRC'].value_counts()

1.0    13781
0.0     1338
2.0      213
3.0        1
Name: P_NUMVRC, dtype: int64

In [50]:
cpox = df[['HAD_CPOX', 'SEX', 'P_NUMVRC']]
cpox = cpox.dropna()
cpox = cpox[cpox['P_NUMVRC']>0]

cpox['P_NUMVRC'].value_counts()

1.0    13781
2.0      213
3.0        1
Name: P_NUMVRC, dtype: int64

In [53]:
cpox['HAD_CPOX'].value_counts()

2     13830
1       121
77       44
Name: HAD_CPOX, dtype: int64

In [55]:
cpox[cpox['HAD_CPOX']==77]

HAD_CPOX  SEX  P_NUMVRC
221          77    2       1.0
222          77    1       1.0
549          77    2       1.0
653          77    1       1.0
1228         77    2       1.0
4132         77    2       1.0
5407         77    1       1.0
5768         77    2       1.0
5819         77    1       1.0
6398         77    1       1.0
6782         77    2       1.0
7241         77    1       1.0
8028         77    2       1.0
8864         77    2       1.0
11728        77    1       1.0
12094        77    2       1.0
13433        77    1       1.0
13556        77    1       1.0
13885        77    1       1.0
14311        77    1       1.0
15491        77    2       1.0
16328        77    1       1.0
16728        77    2       1.0
16854        77    2       1.0
17315        77    1       1.0
17835        77    2       1.0
18515        77    2       1.0
18772        77    2       1.0
18941        77    2       1.0
21063        77    2       1.0
22089        77    1       1.0
22658        77    2       1.0
23063        77    1       1.0
23205        77    2       1.0
24413        77    1       1.0
24461        77    2       1.0
25575        77    2       1.0
25766        77    1       1.0
25775        77    1       1.0
25783        77    1       1.0
25848        77    2       1.0
27471        77    1       1.0
27572        77    1       1.0
27824        77    1       1.0

In [72]:
vm1 = cpox[(cpox.SEX== 1) & (cpox.HAD_CPOX == 1)]
vm2 = cpox[(cpox.SEX== 1) & (cpox.HAD_CPOX == 2)]
male = vm1['P_NUMVRC'].count()/vm2['P_NUMVRC'].count()
print(male)

vf1 = cpox[(cpox.SEX== 2) & (cpox.HAD_CPOX == 1)]
vf2 = cpox[(cpox.SEX== 2) & (cpox.HAD_CPOX == 2)]
female = vf1['P_NUMVRC'].count()/vf2['P_NUMVRC'].count()
print(female)



0.009675583380762664
0.0077918259335489565


In [73]:
vm2['P_NUMVRC'].count()

7028

# D: 

**Correlation**

A correlation is a statistical relationship between two variables. If we wanted to know if vaccines work, we might look at the correlation between the use of the vaccine and whether it results in prevention of the infection or disease [1]. In this question, you are to see if there is a correlation between having had the chicken pox and the number of chickenpox vaccine doses given (varicella).

In [122]:
import pandas as pd
import numpy as np
import scipy.stats as stats
df = pd.read_csv("assignments/assignment2/assets/NISPUF17.csv")

In [149]:
cpox_df = df[['HAD_CPOX', 'P_NUMVRC']]
cpox_df = cpox_df.dropna()
print(cpox_df['P_NUMVRC'].value_counts())
print(cpox_df['HAD_CPOX'].value_counts())



1.0    13781
0.0     1338
2.0      213
3.0        1
Name: P_NUMVRC, dtype: int64
2     15113
1       173
77       47
Name: HAD_CPOX, dtype: int64


In [150]:
#cpox_df = cpox_df[(cpox_df['P_NUMVRC']<3.0) & (cpox_df['P_NUMVRC']>0.0)]
cpox_df= cpox_df[(cpox_df.HAD_CPOX==1) | (cpox_df.HAD_CPOX==2)]
cpox_df['P_NUMVRC'].value_counts()
cpox_df['HAD_CPOX'].value_counts()

2    15113
1      173
Name: HAD_CPOX, dtype: int64

In [156]:
cpox_df

HAD_CPOX  P_NUMVRC
3             2       1.0
4             2       0.0
6             2       1.0
7             2       0.0
8             1       0.0
...         ...       ...
28446         2       1.0
28447         2       1.0
28449         2       1.0
28452         2       1.0
28454         2       1.0

[15286 rows x 2 columns]

In [151]:
corr, pval=stats.pearsonr(cpox_df["HAD_CPOX"],cpox_df["P_NUMVRC"])

In [152]:
corr

0.07044873460147986

In [154]:
pval

2.7780263182916748e-18

In [155]:
np.corrcoef(cpox_df['HAD_CPOX'], cpox_df['P_NUMVRC'])

array([[1.        , 0.07044873],
       [0.07044873, 1.        ]])